### ISSUE TITLE
Admin create of defined compounds via Put

### ISSUE THREAD

**Issue by cmgrulke** [Chemical-Curation/chemcurator_django#35]
**As a** admin
**I want** to be able to create defined compounds using PUT
**So that** I can push in chemicals registered in legacy systems or registered in parrallel deployments of the ChemReg application into this system

**Request Example**
`PUT https://api.chemreg.epa.gov/defined-compounds/DTXCID319949`

**Acceptance Criteria:**
As an admin, I can PUT using any string as a compound id in the url to create that compound.  Other users can only PUT updates to already defined records.  Uniquness checks will be done on the compound id in the url to make sure it doesn't conflict with a compound id in the database.  No uniqueness check will be done on inchikey.

**Test:**

**Comment by mcovalt** [comment]
> As an admin, I can PUT using any string as a compound id in the url to create that compound.

Would it be an acceptable criteria to state that all strings will still follow a somewhat strict pattern?

```regex
^(?<prefix>[a-zA-Z]+)CID\d+$
```
<sup>[Pattern 1 explanation](https://regexr.com/4urau)</sup>

Just not necessarily the pattern stated in previous tickets?

```regex
^(?<prefix>[a-zA-Z]+)CID(?<checksum>\d)0([2-9]\d{6}|\d{8,})$
```
<sup>[Pattern 2 explanation](https://regexr.com/4urcq)</sup>

Or should it match any alphanumeric string?
```regex
^\w+$
```
<sup>[Pattern 3 explanation](https://regexr.com/4urdo)</sup>

Or should it match any string whatsoever?
```regex
^.+$
```
<sup>[Pattern 4 explanation](https://regexr.com/4urdu)</sup>

**Comment by cmgrulke** [comment]
I am not sure exactly what I want in terms of string flexibility.  Since this is limited to an ADMIN, the pool of people who could do this is really limited.  I think we limit it to an alphanumeric string.

**Comment by cmgrulke** [comment]
Also, we should probably limit the size of the string to something under 50 characters.  I am assuming that you guys already have some plan for issuing back error messages when a request is invalid and can craft informative messages for failures in these validation.

**Comment by debboutr** [comment]
based on what I'm reading [here](https://restfulapi.net/rest-put-vs-post/) it seems like if we are creating a resource, we should be using POST and just include the CID in the body?
> I want to be able to create defined compounds using PUT

would it make sense to just remove the inchikey uniqueness validation when the `"cid"` is provided?

**Comment by cmgrulke** [comment]
So the entire purpose of the PUT command is to place a fully defined resource (from the request body) at the request URL.  What I am trying to accomplish here seems like a put, but I am flexible.

**Comment by debboutr** [comment]
so after some capitulation we have decided to use a POST to create resources this way, but it does bring up some questions:
1. should any user be able to mint their own CID when creating a compound, i.e., right now any user can POST a molfile get an auto-generated CID, but if POST'd with a CID, admin rights will be necessary?
2. should the checksum and regex validation used on minted CIDs be applied to the POST'd CID? do we just check that it is unique and that's all?
3. Can a user post a CID in the body with a SMILES/v2000? or should there be a limitation that if the CID is included it can only be v3000? 
4. will we ever want the user to supply the inchikey? right now if one is provided, the one calculated from the molfilev2/3/smiles is what is stored.

**Comment by cmgrulke** [comment]
1. Admin rights are required
2. It should be a unique alpha numeric
3. Only with v3000
4. The user must always provide the inchikey

**Comment by debboutr** [comment]
just to be clear
1. the user will always have to be admin when making a POST to `definedCompounds`?
2. if the `"cid"` is in the body of the POST, then the `"inchikey"` must be provided? or should it be a required field for any POST?

**Comment by cmgrulke** [comment]
1) no admin permission is required to define the cid. 
2) if an admin is posting and defining the cid as part of the request body, they must also provide an inchikey

**Comment by mcovalt** [comment]
Closed in #124
